clusterloader2/pkg/measurement/common/slos/api_responsiveness_prometheus.go

```diff
@@ -0,0 +1,266 @@
+
+	latencyQuery = "quantile_over_time(0.99, apiserver:apiserver_request_latency:histogram_quantile[%v])"
+	countQuery   = "sum(increase(apiserver_request_latency_seconds_count[%v])) by (resource, subresource, scope, verb)"
+
+	latencyWindowSize = 5 * time.Minute
+)
+
+func init() {
+	measurement.Register(apiResponsivenessPrometheusMeasurementName, createAPIResponsivenessPrometheusMeasurement)
+}
+
+func createAPIResponsivenessPrometheusMeasurement() measurement.Measurement {
+	return &apiResponsivenessMeasurementPrometheus{}
+}
+
+type apiResponsivenessMeasurementPrometheus struct {
+	startTime time.Time
+	apiCalls  map[string]*apiCall
+}
+
+func (a *apiResponsivenessMeasurementPrometheus) Execute(config *measurement.MeasurementConfig) ([]measurement.Summary, error) {
+	var summaries []measurement.Summary
+
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return summaries, err
+	}
+
+	switch action {
+	case "start":
+		a.start()
+	case "gather":
+		summary, err := a.gather(config.ClientSets.GetClient())
+		if err == nil || errors.IsMetricViolationError(err) {
+			summaries = append(summaries, summary)
+		}
+		return summaries, err
+	default:
+		return summaries, fmt.Errorf("unknown action %v", action)
+	}
+
+	return summaries, nil
+}
+
+// Dispose cleans up after the measurement.
+func (a *apiResponsivenessMeasurementPrometheus) Dispose() {}
+
+// String returns string representation of this measurement.
+func (*apiResponsivenessMeasurementPrometheus) String() string {
+	return apiResponsivenessPrometheusMeasurementName
+}
+
+func (a *apiResponsivenessMeasurementPrometheus) start() {
+	a.startTime = time.Now()
+}
+
+func (a *apiResponsivenessMeasurementPrometheus) gather(c clientset.Interface) (measurement.Summary, error) {
+	apiCalls, err := a.gatherApiCalls(c)
+	if err != nil {
+		klog.Errorf("%s: samples gathering error: %v", a, err)
+	}
+
+	metrics := &apiResponsiveness{ApiCalls: apiCalls}
+	sort.Sort(sort.Reverse(metrics))
+	var badMetrics []string
+	top := 5
+	for i := range metrics.ApiCalls {
+		isBad := false
+		latencyThreshold := getLatencyThreshold(&metrics.ApiCalls[i])
+		if metrics.ApiCalls[i].Latency.Perc99 > latencyThreshold {
+			isBad = true
+			badMetrics = append(badMetrics, fmt.Sprintf("got: %+v; expected perc99 <= %v", metrics.ApiCalls[i], latencyThreshold))
+		}
+		if top > 0 || isBad {
+			top--
+			prefix := ""
+			if isBad {
+				prefix = "WARNING "
+			}
+			klog.Infof("%s: %vTop latency metric: %+v; threshold: %v", a, prefix, metrics.ApiCalls[i], latencyThreshold)
+		}
+	}
+	if len(badMetrics) > 0 {
+		return metrics, errors.NewMetricViolationError("top latency metric", fmt.Sprintf("there should be no high-latency requests, but: %v", badMetrics))
+	}
+	return metrics, nil
+}
+
+func (a *apiResponsivenessMeasurementPrometheus) gatherApiCalls(c clientset.Interface) ([]apiCall, error) {
+	measuermentDuration := time.Since(a.startTime)
```

Comment:
We should store time.Now in a variable and use it in both queries

Reply:
I'm not sure what you mean. Let's discuss this offline.

Reply:
Discussed in person. 

The time can be set by specifying the time parameter in the query, e.g.

/api/v1/query?query=<query>&time=2015-07-01T20:10:51.781Z

Reply:
Done.